### DISCUSSION TITLE
Expectations and best practices for an NVIDIA Tesla P40

### DISCUSSION THREAD

**Discussion by andrewreid** [blakeblackshear/frigate#8581] [Ask A Question] [answered]
For a while, I've run Frigate on a little i5-6500T HP micro PC, using OpenVINO to great effect. With only two cameras, it was working well enough on 0.13-beta4, but the old iGPU is a bit gutless and I've been experimenting with a NVIDIA Tesla P40 I picked up, with an eye on capacity for more cameras.

Now, running two cameras, the i5 barely broke a sweat. That's my "baseline". I've got the Tesla P40 in an ESXi host passing through an 8GB vGPU to the Frigate VM. `nvidia-smi` shows the three processes – `frigate.detector.tensorrt` and two `ffmpeg` instances – using a combined 1366MiB of video memory.

I'm now trying to answer two questions: 1) why is `frigate.detector.tensorrt` using 100% of a CPU, fairly well constantly and 2) why is the GPU usage > 90% most of the time?

My [config.yml](https://gist.github.com/andrewreid/eaf483ee9286b7d7443e296697250c5c) is setup with `go2rtc` pulling in the two cameras, then transcoding the main feed into a 1280x720 stream for `detect`. My [docker-compose.yml](https://gist.github.com/andrewreid/54450af62337d0a85470fe47fe62011e) is setup to generate a `yolov7-640` model.

Looking at the process output on the Frigate VM, I've only got two accelerated `ffmpeg` processes running:

```
 320819 ?        Sl     1:41 ffmpeg -hide_banner -v error -hwaccel cuda -hwaccel_output_format cuda -fflags nobuffer -flags low_delay -timeout 5000000 -user_agent go2rtc/ffmpeg -rtsp_flags prefer_tcp -i rtsp://127.0.0.1:8554/shed-1?video -fpsmax 5 -c:v h264_nvenc -g 50 -bf 0 -profile:v high -level:v auto -preset:v p2 -tune:v ll -an -vf scale_cuda=1280:720 -user_agent ffmpeg/go2rtc -rtsp_transport tcp -f rtsp rtsp://127.0.0.1:8554/dffca415896752349fc5e50b0e3f527f
 320820 ?        Sl     1:20 ffmpeg -hide_banner -v error -hwaccel cuda -hwaccel_output_format cuda -fflags nobuffer -flags low_delay -timeout 5000000 -user_agent go2rtc/ffmpeg -rtsp_flags prefer_tcp -i rtsp://127.0.0.1:8554/front_lawn?video -fpsmax 5 -c:v h264_nvenc -g 50 -bf 0 -profile:v high -level:v auto -preset:v p2 -tune:v ll -an -vf scale_cuda=1280:720 -user_agent ffmpeg/go2rtc -rtsp_transport tcp -f rtsp rtsp://127.0.0.1:8554/bc91ef19ff9f3f7bc98ac06092f8a50f
```

I'm trying to understand why my (admittedly older) datacenter-grade GPU is being maxed out by what I thought would be a fairly modest workload. Can anyone see anything I'm not doing right? And why would CPU usage be so high on `frigate.detector.tensorrt`? Grateful for any insights.

**Comment by NickM-27** [comment]
There are 2 big issues here:
1. You are not using the GPU for decoding the streams when they come in to frigate, so this is going to use a decent amount of cpu
2. you are running transcoding for a stream that as far as I can tell will only be used in frigate. There is no reason to do this, frigate can already scale the stream while decoding instead which reduces the load because no encoding (and then subsequent decoding) is required.

Then a smaller issue is that you are running on a 640x640 model which is much larger than openvino, I'd suggest using the yolo 320 model that is shown in the docs

**Reply by andrewreid** [reply]
I've made some changes and I'm seeing improvements (updated [config.yml](https://gist.github.com/andrewreid/a81c834ea6dee4e510d273fa42641c25)):

1. Top-level `ffmpeg` with `hwaccel_args: preset-nvidia-h264` 
2. `go2rtc` streams pulls in one full-quality stream per camera
3. Each camera in `config.yml` now has both `detect` and `record` roles against the same input 
4. Changed model to YOLOv7-320

Improvements: inference speed is way down to 10.5ms and GPU usage is next to nothing (sub 5%). 

What hasn't changed is what seems to be largely unchanged is the large CPU usage on the `detect` processes. Frigate's system dashboard shows high usage that accords with output from `htop`:

<img width="1229" alt="image" src="https://github.com/blakeblackshear/frigate/assets/114979/3ce35ab6-2b4d-4ee7-a9f9-ae043a58e98f">

Digging through `htop` and process trees it looks like the `frigate.process:shed-1` and `frigate.process:front_lawn` are the two CPU-hungry processes. `frigate.detector.tensorrt` isn't using much at all – maybe 10%ish.

My gut feeling is that I'm seeing too much CPU usage for just two cameras. What's your take?

**Reply by NickM-27** [reply, marked as answer]
Check your motion detection, I bet it's too sensitive currently. 0.13 motion detection is more sensitive by default. https://deploy-preview-6262--frigate-docs.netlify.app/configuration/motion_detection walks through how to tune this.

**Reply by andrewreid** [reply]
Tuned the motion detection, masked out a couple of trees... We're cooking with gas, now! Cheers, Nick.
